aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -87,18 +168,105 @@ public void close() throws IOException {
 
     super.close();
     closed = true;
+    currentStagingFile = null;
 
     try {
       stream.close();
 
-      s3.putObject(
-          PutObjectRequest.builder().bucket(location.bucket()).key(location.key()).build(),
-          RequestBody.fromFile(stagingFile));
+      completeUploads();
     } finally {
-      if (!stagingFile.delete()) {
-        LOG.warn("Could not delete temporary file: {}", stagingFile);
+      stagingFiles.forEach(f -> {
```

Comment:
Can we also delete the temp file progressively instead of deleting everything after close() is called?

Reply:
Good catch! That was the intent and I think at some point I had it in the async upload task.  Should be easy to add back there after the upload is successful.

Reply:
Also, we can add reliability by using `Tasks.foreach`. That would allow retries, failure logging, and deleting each file even if another has failed.

Reply:
Yeah, let me try switching over to Tasks.  I was having trouble with figuring out how to properly handle the exception cases with CompletableFuture without swallowing the failures.

Reply:
@rdblue looking at `Tasks.foreach`, I'm not sure how well it fits in this case.  It looks like `Tasks` assumes that you know the full units of work and will submit them in parallel and block.  In this case, we are progressively adding more work and then need to block across all of those at the end to complete the multipart upload.  Now, I could wrap the `Tasks.foreach` call in future, but that seems like we're just wrapping one async framework with another.

I think we can handle the concerns of deleting the files, aborting the upload, and logging with `CompletableFuture` framework and retries should be left to the S3Client (trying to avoid retries on top of retries).

Thoughts?

Reply:
Are you saying that `stagingFiles` is changing as this loops through it?

I think that when we try to revert the uploads, we should know that no more uploads are happening concurrently so that the tasks can run. Maybe I'm misunderstanding your point though?

Reply:
I think I'm the one who got confused here.  I though you were suggesting to move a different section to using `Tasks`.  For the actual cleanup, Tasks will probably work.  I'll take another look.

Reply:
Ok, so there are actually two areas to resolve here (which was my original confusion).

1.  Delete the staging file when an upload is success full, which I added
2.  Convert the cleanup during abort to use Tasks framework.